From 7b8256f72e427c5f42e0d70110061e03c06e8e74 Mon Sep 17 00:00:00 2001
From: DavidZalman101 <davidzalman.101@gmail.com>
Date: Tue, 24 Jun 2025 21:43:10 +0300
Subject: [PATCH 02/12] lab6 - Exercise 1.

---
 kern/syscall.c | 6 +++++-
 kern/trap.c    | 2 ++
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/kern/syscall.c b/kern/syscall.c
index e749d2b..22bf280 100644
--- a/kern/syscall.c
+++ b/kern/syscall.c
@@ -557,7 +557,7 @@ static int
 sys_time_msec(void)
 {
 	// LAB 6: Your code here.
-	panic("sys_time_msec not implemented");
+	return time_msec();
 }
 
 // Dispatches to the correct kernel function, passing the arguments.
@@ -628,6 +628,10 @@ syscall(uint32_t syscallno, uint32_t a1, uint32_t a2, uint32_t a3, uint32_t a4,
 			r_val = sys_env_set_trapframe((envid_t) a1, (struct Trapframe*) a2);
 			goto done;
 
+		case SYS_time_msec:
+			r_val = sys_time_msec();
+			goto done;
+
 		default:
 			r_val = -E_INVAL;
 	}
diff --git a/kern/trap.c b/kern/trap.c
index 89ec697..9bae126 100644
--- a/kern/trap.c
+++ b/kern/trap.c
@@ -255,6 +255,8 @@ trap_dispatch(struct Trapframe *tf)
 
 		// clock interrupt
 		case IRQ_OFFSET	+ IRQ_TIMER:
+			if (thiscpu->cpu_id == 0)
+				time_tick();
 			lapic_eoi();
 			sched_yield(); // never returns
 
-- 
1.9.1

